### PR TITLE
feat: Add EventSubmitter to Pinpoint

### PR DESCRIPTION
**Pull request by sdhuka**
*Issue #, if available:*

*Description of changes:*
feat: Add EventSubmitter to Pinpoint

*How did you test these changes?*
- Manually tested

- [ *] Added Unit Tests

*Documentation update required?*
- [* ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginBehavior.kt [hunk 1]
```diff
@@ -50,19 +57,18 @@ internal class AWSPinpointAnalyticsPluginBehavior(
...
-                    attributes[key] = (analyticsProperty.value as AnalyticsStringProperty).value
+                    attributes[key] = analyticsProperty.value
                 }
                 is AnalyticsBooleanProperty -> {
-                    attributes[key] = (analyticsProperty.value as AnalyticsBooleanProperty).value.toString()
+                    attributes[key] = analyticsProperty.value.toString()
                 }
                 is AnalyticsIntegerProperty -> {
-                    metrics[key] = (analyticsProperty.value as AnalyticsIntegerProperty).value.toDouble()
+                    metrics[key] = analyticsProperty.value.toDouble()
```
Comment (tjleing) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginBehavior.kt:68]: is `AnalyticsIntegerProperty` and `AnalyticsDoubleProperty` logic swapped here?  just looks weird to have `toDouble()` under `IntegerProperty`

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginKt.kt [hunk 2]
```diff
@@ -24,22 +25,24 @@ import org.json.JSONObject
 /**
```
Comment (tjleing) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginKt.kt:25]: should javadoc be directly above the class definition instead?

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/EventRecorder.kt [hunk 3]
```diff
@@ -15,29 +15,261 @@
...
     val context: Context,
-    val pinpointClient: PinpointClient,
+    private val pinpointClient: PinpointClient,
     private val pinpointDatabase: PinpointDatabase,
-    private val coroutineDispatcher: CoroutineDispatcher = Dispatchers.Default
+    private val targetingClient: TargetingClient,
+    private val coroutineDispatcher: CoroutineDispatcher = Dispatchers.Default,
+    private val logger: Logger =
+        Amplify.Logging.forNamespace(
+            AWS_PINPOINT_ANALYTICS_LOG_NAMESPACE.format(EventRecorder::class.java.simpleName)
```
Comment (tjleing) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/EventRecorder.kt:53]: looks more specific than the generalized way I've been doing it in my code -- I'll implement this as well (fast-follow PR with unit tests?)

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/PinpointManager.kt [hunk 4]
```diff
@@ -19,37 +19,57 @@ import android.telephony.TelephonyManager
...
+        val androidAppDetails = AndroidAppDetails(context, awsPinpointConfiguration.appId)
+        val androidDeviceDetails = AndroidDeviceDetails(getCarrier(context))
+        targetingClient = TargetingClient(
+            pinpointClient,
+            PinpointNotificationClient(),
+            SharedPrefsUniqueIdService(sharedPrefs),
+            sharedPrefs,
+            androidAppDetails,
+            androidDeviceDetails,
+            context
```
Comment (tjleing) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/PinpointManager.kt:62]: should this constructor be called with `=` syntax like in `analyticsClient` below?

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/models/PinpointEvent.kt [hunk 5]
```diff
@@ -14,32 +14,42 @@
...
+    companion object {
+        fun fromJsonString(jsonString: String): PinpointEvent {
+            return Json.decodeFromString<PinpointEvent>(jsonString)
+        }
+    }
+
     fun toJsonString(): String {
-        val gson = Gson()
-        return gson.toJson(this)
+        return Json.encodeToString(this)
```
Comment (tjleing) [inline aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/models/PinpointEvent.kt:53]: shouldn't this be autogenerated by `@Serializable` -- something like `Json.encodeToString(PinpointEvent.serializer(), event)`?  or are these for better syntax

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginBehavior.kt (hunk 1 above)
Comment (sdhuka) [reply aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginBehavior.kt:68]: metric is a map of double so we have to convert it.

https://github.com/aws-amplify/amplify-android/blob/main/aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPlugin.java#L217

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginKt.kt (hunk 2 above)
Comment (sdhuka) [reply aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginKt.kt:25]: I have not seen complaint from lint regarding it.

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginBehavior.kt (hunk 1 above)
Comment (tjleing) [reply aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AWSPinpointAnalyticsPluginBehavior.kt:68]: ah of course, thanks!

**Pull request closed without merging**